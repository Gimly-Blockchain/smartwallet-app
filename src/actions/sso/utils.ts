import { IssuerPublicProfileSummary } from './types'
import { Identity } from 'jolocom-lib/js/identity/identity'

export const parsePublicProfile = (identity: Identity) => {
  const { publicProfile, did } = identity
  if (!publicProfile) {
    return {
      did,
    }
  }
  const { id, ...parsedProfile } = publicProfile.claim
  return {
    did,
    publicProfile: parsedProfile as IssuerPublicProfileSummary,
  }
}
