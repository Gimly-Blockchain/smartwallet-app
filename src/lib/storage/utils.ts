// TODO rework
// Multiple credentials of the same type, i.e. ['Credential', 'ProofOfNameCredential']
import {
  CredentialEntity,
  VerifiableCredentialEntity,
} from './entities'

interface ModifiedCredentialEntity {
  propertyName: string
  propertyValue: string[]
  verifiableCredential: VerifiableCredentialEntity
}

export const groupAttributesByCredentialId = (
  credentials: CredentialEntity[],
): ModifiedCredentialEntity[] => {
  // Convert values to arrays for easier concatination later
  const modifiedAttributes = credentials.map(credential => ({
    ...credential,
    propertyValue: [credential.propertyValue], // "eugeniu@jolocom.com" -> ["eugeniu@jolocom.com"]
  }))

  // Helper function
  const findByCredId = (
    arrToSearch: ModifiedCredentialEntity[],
    value: ModifiedCredentialEntity,
  ) =>
    arrToSearch.findIndex(
      entry => entry.verifiableCredential.id === value.verifiableCredential.id,
    )

  debugger
  return modifiedAttributes.reduce(
    (acc: ModifiedCredentialEntity[], curr: ModifiedCredentialEntity) => {
      const matchingIndex = findByCredId(acc, curr)

      // This seems to be a hack, uniting individual claims from the credential back together
      if (matchingIndex >= 0) {
        acc[matchingIndex].propertyValue = [
          ...acc[matchingIndex].propertyValue,
          ...curr.propertyValue,
        ]
        return acc
      } else {
        return [...acc, curr]
      }
    },
    [],
  )
}
