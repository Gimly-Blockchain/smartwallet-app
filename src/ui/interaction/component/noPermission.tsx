import React, { FC } from 'react'
import {
  Platform,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native'
import I18n from '../../../locales/i18n'
import strings from '../../../locales/strings'
import { Colors, Spacing } from '../../../styles'
import {
  centeredText,
  fontMain,
  textSM,
  textXL,
  textXS,
} from '../../../styles/typography'

const styles = StyleSheet.create({
  notAuthorizedOverlay: {
    flex: 1,
    backgroundColor: Colors.black065,
    height: '100%',
    width: '100%',
    alignItems: 'center',
    zIndex: 1,
    justifyContent: 'center',
  },
  scanText: {
    color: Colors.sandLight,
    fontSize: textXL,
    fontFamily: fontMain,
    ...centeredText,
  },
  notAuthorizedDescription: {
    color: Colors.sandLight080,
    fontSize: textSM,
    paddingTop: Spacing.XS,
    paddingHorizontal: Spacing.XXL,
    marginBottom: Spacing.XXL,
    fontFamily: fontMain,
    lineHeight: 24,
    ...centeredText,
  },
  enableButtonText: {
    fontSize: textXS,
    fontFamily: fontMain,
    lineHeight: 20,
    color: Platform.select({
      android: Colors.white,
      ios: Colors.nativeIosBlue,
    }),
    textDecorationLine: Platform.select({
      android: 'underline',
      ios: 'none',
    }),
    ...centeredText,
  },
  enableButtonWrapper: {
    padding: 10,
  },
})

interface Props {
  onPressEnable: () => void
}

export const NoPermissionComponent: FC<Props> = (props: Props) => {
  return (
    <React.Fragment>
      <View style={styles.notAuthorizedOverlay}>
        <Text style={styles.scanText}>{I18n.t(strings.SCAN_QR)}</Text>
        <Text style={styles.notAuthorizedDescription}>
          {I18n.t(
            strings.ENABLE_ACCESS_SO_YOU_CAN_START_TAKING_PHOTOS_AND_VIDEOS,
          )}
        </Text>
        <TouchableOpacity
          style={styles.enableButtonWrapper}
          onPress={props.onPressEnable}
        >
          <Text style={styles.enableButtonText}>
            {I18n.t(strings.ENABLE_CAMERA_ACCESS)}
          </Text>
        </TouchableOpacity>
      </View>
      <View style={styles.notAuthorizedOverlay} />
    </React.Fragment>
  )
}
