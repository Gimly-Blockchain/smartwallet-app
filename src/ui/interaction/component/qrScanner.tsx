import React from 'react'
import {
  StyleSheet,
  Text,
  Platform,
  View,
  Dimensions,
  TouchableHighlight,
  AppState,
  AppStateStatus,
  PermissionStatus,
} from 'react-native'
import { QrScanEvent } from 'src/ui/generic/qrcodeScanner'
import { NavigationScreenProps } from 'react-navigation'
import { Colors, Spacing } from '../../../styles'
import { black065, white } from '../../../styles/colors'
import {
  centeredText,
  fontMain,
  textSM,
  textXL,
  textXS,
} from '../../../styles/typography'
import I18n from 'src/locales/i18n'
import strings from 'src/locales/strings'
import { RNCamera } from 'react-native-camera'
import { TorchOffIcon, TorchOnIcon } from '../../../resources'
import { PermissionsAndroid } from 'react-native'
import AndroidOpenSettings from 'react-native-android-open-settings'

const QRScanner = require('react-native-qrcode-scanner').default

export interface QrScanEvent {
  data: string
}

interface Props extends NavigationScreenProps {
  onScannerSuccess: (e: QrScanEvent) => void
}

interface State {
  isTorch: boolean
  key: number
  permission: PermissionStatus
  isCamera: boolean
}

const SCREEN_HEIGHT = Dimensions.get('window').height
const SCREEN_WIDTH = Dimensions.get('window').width
const MARKER_SIZE = SCREEN_WIDTH * 0.75

const styles = StyleSheet.create({
  rectangle: {
    height: MARKER_SIZE,
    width: MARKER_SIZE,
    borderRadius: 2,
    borderWidth: 2,
    borderColor: white,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'transparent',
  },
  bottomOverlay: {
    flex: 1,
    backgroundColor: Colors.black065,
    height: '100%',
    width: '100%',
    alignItems: 'center',
    zIndex: 1,
    justifyContent: 'space-between',
    paddingBottom: '6%',
  },
  notAuthorizedOverlay: {
    flex: 1,
    backgroundColor: Colors.black065,
    height: '100%',
    width: '100%',
    alignItems: 'center',
    zIndex: 1,
    justifyContent: 'center',
  },
  horizontalOverlay: {
    height: MARKER_SIZE,
    width: SCREEN_WIDTH,
    backgroundColor: black065,
  },
  descriptionText: {
    marginTop: Spacing.SM,
    color: Colors.sandLight,
    fontSize: textSM,
    paddingHorizontal: Spacing['3XL'],
    fontFamily: fontMain,
    lineHeight: 20,
    ...centeredText,
  },
  scanText: {
    color: Colors.sandLight,
    fontSize: textXL,
    fontFamily: fontMain,
    ...centeredText,
  },
  notAuthorizedDescription: {
    color: Colors.sandLight080,
    fontSize: textSM,
    paddingTop: Spacing.XS,
    paddingHorizontal: Spacing.XXL,
    fontFamily: fontMain,
    lineHeight: 24,
    ...centeredText,
  },
  permissionButton: {
    color: Colors.white,
    fontSize: textXS,
    textDecorationLine: 'underline',
    fontFamily: fontMain,
    lineHeight: 20,
    ...centeredText,
  },
  torch: {
    width: 80,
    height: 80,
    alignItems: 'center',
    justifyContent: 'center',
  },
})

export class QRCodeScanner extends React.Component<Props, State> {
  private scanner: typeof QRScanner
  private removeFocusListener: (() => void) | undefined

  public constructor(props: Props) {
    super(props)
    this.state = {
      isTorch: false,
      // NOTE: the key is used to invalidate the previously rendered component as
      // we need to rerender and remount the camera to ensure it is properly setup
      key: Date.now(),
      permission: PermissionsAndroid.RESULTS.DENIED,
      isCamera: false,
    }
  }

  public async componentDidMount(): Promise<void> {
    // NOTE: timeout the rendering of the camera to avoid the screen stuttering
    setTimeout(() => this.setState({ isCamera: true }), 200)
    const permission = await this.requestCameraPermission()
    if (this.props.navigation) {
      this.removeFocusListener = this.props.navigation.addListener(
        'willFocus',
        () => {
          if (permission === PermissionsAndroid.RESULTS.GRANTED) {
            this.scanner.reactivate()
            // NOTE: the re-render and the re-mount should only fire during the willFocus event
            this.setState({ key: Date.now() })
            // NOTE: force an update to force remounting of the Camera
            this.forceUpdate()
          }
        },
      ).remove
    }
  }

  public componentWillUnmount(): void {
    if (this.removeFocusListener) this.removeFocusListener()
  }

  private requestCameraPermission = async () => {
    const permission = await PermissionsAndroid.request(
      PermissionsAndroid.PERMISSIONS.CAMERA,
    )

    this.setState({
      permission,
    })

    return permission
  }

  // detect when the focus is back on the app screen after navigating
  // to settings, in order to check if the permissions changed
  private promiseOpenSettings = (cb: () => void) => {
    return new Promise((resolve, reject) => {
      const listener = (state: AppStateStatus) => {
        if (state === 'active') {
          AppState.removeEventListener('change', listener)
          resolve()
        }
      }
      AppState.addEventListener('change', listener)
      try {
        cb()
      } catch (e) {
        AppState.removeEventListener('change', listener)
        reject(e)
      }
    })
  }

  private openSettings = () => {
    if (Platform.OS === 'ios') {
      // navigate to iOS settings through DeepLink
    } else {
      this.promiseOpenSettings(AndroidOpenSettings.appDetailsSettings).then(
        this.requestCameraPermission,
      )
    }
  }

  private onTorchChange = (state: boolean) => {
    this.setState({
      isTorch: state,
    })
  }

  private renderCamera = () => {
    const { onScannerSuccess } = this.props
    const cameraProps = {
      key: this.state.key,
      captureAudio: false,
      flashMode: this.state.isTorch
        ? RNCamera.Constants.FlashMode.torch
        : RNCamera.Constants.FlashMode.off,
    }
    return (
      <React.Fragment>
        {/* NOTE: The QRScanner is positioned as absolute to
         allow the interaction menu on top of it */}
        {this.state.isCamera ? (
          <QRScanner
            containerStyle={{ position: 'absolute' }}
            cameraProps={cameraProps}
            ref={(ref: React.Component) => (this.scanner = ref)}
            fadeIn
            onRead={onScannerSuccess}
            cameraStyle={{ height: SCREEN_HEIGHT }}
          />
        ) : null}
        <View style={{ flexDirection: 'row' }}>
          <View style={styles.horizontalOverlay} />
          <View style={styles.rectangle} />
          <View style={styles.horizontalOverlay} />
        </View>
        <View style={styles.bottomOverlay}>
          <Text style={styles.descriptionText}>
            {I18n.t(
              strings.ITS_ALL_AUTOMATIC_JUST_PLACE_YOUR_PHONE_ABOVE_THE_CODE,
            )}
          </Text>

          <TouchableHighlight
            onPressIn={() => this.onTorchChange(true)}
            onPressOut={() => this.onTorchChange(false)}
            activeOpacity={1}
            underlayColor={'transparent'}
            style={styles.torch}
          >
            {this.state.isTorch ? <TorchOnIcon /> : <TorchOffIcon />}
          </TouchableHighlight>
        </View>
      </React.Fragment>
    )
  }

  private renderNoPermissionView = () => {
    return (
      <React.Fragment>
        <View style={styles.notAuthorizedOverlay}>
          <Text style={styles.scanText}>{I18n.t(strings.SCAN_QR)}</Text>
          <Text style={styles.notAuthorizedDescription}>
            {I18n.t(
              strings.ENABLE_ACCESS_SO_YOU_CAN_START_TAKING_PHOTOS_AND_VIDEOS,
            )}
          </Text>
        </View>
        <View style={styles.notAuthorizedOverlay}>
          <Text
            style={styles.permissionButton}
            onPress={async () => {
              if (
                this.state.permission ===
                PermissionsAndroid.RESULTS.NEVER_ASK_AGAIN
              ) {
                this.openSettings()
              } else {
                await this.requestCameraPermission()
              }
            }}
          >
            {I18n.t(strings.ENABLE_CAMERA_ACCESS)}
          </Text>
        </View>
      </React.Fragment>
    )
  }

  public render() {
    return this.state.permission === PermissionsAndroid.RESULTS.GRANTED
      ? this.renderCamera()
      : this.renderNoPermissionView()
  }
}
