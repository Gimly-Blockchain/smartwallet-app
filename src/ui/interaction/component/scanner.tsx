import React from 'react'
import QRScanner, { Event } from 'react-native-qrcode-scanner'
import { NavigationScreenProps } from 'react-navigation'
import { RNCamera } from 'react-native-camera'
import {
  Dimensions,
  StyleSheet,
  Text,
  TouchableHighlight,
  View,
} from 'react-native'
import I18n from '../../../locales/i18n'
import strings from '../../../locales/strings'
import { TorchOffIcon, TorchOnIcon } from '../../../resources'
import { black065, white } from '../../../styles/colors'
import { Colors, Spacing } from '../../../styles'
import { centeredText, fontMain, textSM } from '../../../styles/typography'

const SCREEN_HEIGHT = Dimensions.get('window').height
const SCREEN_WIDTH = Dimensions.get('window').width
const MARKER_SIZE = SCREEN_WIDTH * 0.75

const styles = StyleSheet.create({
  rectangle: {
    height: MARKER_SIZE,
    width: MARKER_SIZE,
    borderRadius: 2,
    borderWidth: 2,
    borderColor: white,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'transparent',
  },
  bottomOverlay: {
    flex: 1,
    backgroundColor: Colors.black065,
    height: '100%',
    width: '100%',
    alignItems: 'center',
    zIndex: 1,
    justifyContent: 'space-between',
    paddingBottom: '6%',
  },
  horizontalOverlay: {
    height: MARKER_SIZE,
    width: SCREEN_WIDTH,
    backgroundColor: black065,
  },
  descriptionText: {
    marginTop: Spacing.SM,
    color: Colors.sandLight,
    fontSize: textSM,
    paddingHorizontal: Spacing['3XL'],
    fontFamily: fontMain,
    lineHeight: 20,
    ...centeredText,
  },
  torch: {
    width: 80,
    height: 80,
    alignItems: 'center',
    justifyContent: 'center',
  },
})

interface Props extends NavigationScreenProps {
  onScannerSuccess: (jwt: string) => void
}

interface State {
  isTorch: boolean
  reRenderKey: number
  isCameraReady: boolean
}

export class ScannerComponent extends React.Component<Props, State> {
  private scanner!: QRScanner
  private removeFocusListener: (() => void) | undefined

  public constructor(props: Props) {
    super(props)
    this.state = {
      isTorch: false,
      // NOTE: the key is used to invalidate the previously rendered component as
      // we need to rerender and remount the camera to ensure it is properly setup
      reRenderKey: Date.now(),
      isCameraReady: false,
    }
  }

  public async componentDidMount(): Promise<void> {
    // NOTE: timeout the rendering of the camera to avoid the screen stuttering
    setTimeout(() => this.setState({ isCameraReady: true }), 200)
    if (this.props.navigation) {
      this.removeFocusListener = this.props.navigation.addListener(
        'willFocus',
        () => {
          if(this.state.isCameraReady) {
            this.scanner.reactivate()
            // NOTE: the re-render and the re-mount should only fire during the willFocus event
            this.setState({ reRenderKey: Date.now() })
            // NOTE: force an update to force remounting of the Camera
            this.forceUpdate()
          }

        },
      ).remove
    }
  }

  public componentWillUnmount(): void {
    if (this.removeFocusListener) this.removeFocusListener()
  }

  private onTorchChange = (state: boolean) => {
    this.setState({
      isTorch: state,
    })
  }

  public render() {
    const { onScannerSuccess } = this.props
    const cameraProps = {
      key: this.state.reRenderKey,
      captureAudio: false,
      flashMode: this.state.isTorch
        ? RNCamera.Constants.FlashMode.torch
        : RNCamera.Constants.FlashMode.off,
    }

    return (
      <React.Fragment>
        {this.state.isCameraReady && (
          <QRScanner
            //@ts-ignore - see https://github.com/DefinitelyTyped/DefinitelyTyped/issues/29651
            containerStyle={{ position: 'absolute' }}
            cameraProps={cameraProps}
            ref={(ref: QRScanner) => (this.scanner = ref)}
            fadeIn
            onRead={(e: Event) => onScannerSuccess(e.data)}
            //@ts-ignore
            cameraStyle={StyleSheet.create({ height: SCREEN_HEIGHT })}
          />
        )}

        <View style={{ flexDirection: 'row' }}>
          <View style={styles.horizontalOverlay} />
          <View style={styles.rectangle} />
          <View style={styles.horizontalOverlay} />
        </View>
        <View style={styles.bottomOverlay}>
          <Text style={styles.descriptionText}>
            {I18n.t(
              strings.ITS_ALL_AUTOMATIC_JUST_PLACE_YOUR_PHONE_ABOVE_THE_CODE,
            )}
          </Text>
          <TouchableHighlight
            onPressIn={() => this.onTorchChange(true)}
            onPressOut={() => this.onTorchChange(false)}
            activeOpacity={1}
            underlayColor={'transparent'}
            style={styles.torch}
          >
            {this.state.isTorch ? <TorchOnIcon /> : <TorchOffIcon />}
          </TouchableHighlight>
        </View>
      </React.Fragment>
    )
  }
}
