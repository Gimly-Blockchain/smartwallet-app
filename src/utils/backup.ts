import { SoftwareKeyProvider } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'
// @ts-ignore
import eccrypto from 'eccrypto'
import { ICredentialAttrs } from 'jolocom-lib/js/credentials/credential/types'
import { IKeyDerivationArgs } from 'jolocom-lib/js/vaultedKeyProvider/types'

export interface BackupFile {
  keys: EncryptedKey[]
  data: string
}

export interface EncryptedKey {
  pubKey: string
  cipher: string
}

export async function createBackup(
  did: string,
  credentials: ICredentialAttrs[],
  vault: SoftwareKeyProvider,
  derivationArgs: IKeyDerivationArgs,
): Promise<string> {
  const data = {
    did,
    credentials,
    // TODO add more data here like e.g. interaction data
  }
  const publicKey = vault.getPublicKey(derivationArgs)

  const symKey = SoftwareKeyProvider.getRandom(128)
  // @ts-ignore private
  const encryptedData: Buffer = SoftwareKeyProvider.encrypt(
    symKey,
    JSON.stringify(data),
  )
  const encryptedKey = await eccrypto.encrypt(publicKey, Buffer.from(symKey))
  const backupFile: BackupFile = {
    keys: [
      {
        cipher: stringifyEncryptedData(encryptedKey),
        pubKey: publicKey.toString('hex'),
      },
    ],
    data: encryptedData.toString('hex'),
  }
  return JSON.stringify(backupFile)
}

export async function decryptBackup(
  backupFile: BackupFile,
  vault: SoftwareKeyProvider,
  derivationArg: IKeyDerivationArgs,
): Promise<string> {
  const publicKey = vault.getPublicKey(derivationArg)
  const privateKey = vault.getPrivateKey(derivationArg)
  // find encrypted key
  const encryptedKey = backupFile.keys.find(
    key => key.pubKey === publicKey.toString('hex'),
  )
  if (!encryptedKey) throw new Error('Not encrypted for these keys')
  const key = await eccrypto.decrypt(
    privateKey,
    parseEncryptedData(encryptedKey.cipher),
  )
  // @ts-ignore private
  return SoftwareKeyProvider.decrypt(
    key,
    Buffer.from(backupFile.data, 'hex'),
  ).toString()
}

function stringifyEncryptedData(data: {
  iv: Buffer
  ephemPublicKey: Buffer
  ciphertext: Buffer
  mac: Buffer
}): string {
  let hexData = {}
  Object.keys(data).forEach(key => (hexData[key] = data[key].toString('hex')))
  return JSON.stringify(hexData)
}

function parseEncryptedData(data: string): object {
  const hexData = JSON.parse(data)
  let bufferData = {}
  Object.keys(hexData).forEach(
    key => (bufferData[key] = Buffer.from(hexData[key], 'hex')),
  )
  return bufferData
}
