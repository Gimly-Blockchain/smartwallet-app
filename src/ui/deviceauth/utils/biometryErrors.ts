import { Alert, Linking } from 'react-native'

import strings from '../../../locales/strings'
import { BiometryTypes } from '../types'
import { getBiometryIsDisabledText } from '../utils/getText'

export const handleNotEnrolled = (biometryType: BiometryTypes) => {
  Alert.alert(
    getBiometryIsDisabledText(biometryType),
    strings.TO_USE_BIOMETRICS_ENABLE,
    [
      {
        text: strings.SETTINGS,
        onPress: () => Linking.openSettings(),
      },
      {
        text: strings.CANCEL,
        // onPress: () => console.log('Cancel Pressed'),
        style: 'cancel',
      },
    ],
    { cancelable: false },
  )
}
