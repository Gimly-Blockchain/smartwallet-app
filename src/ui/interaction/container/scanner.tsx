import React, { useEffect, useState } from 'react'
import { Animated, AppState, AppStateStatus, Platform } from 'react-native'
import {
  NavigationEventSubscription,
  NavigationScreenProps,
} from 'react-navigation'
/* TODO: When using the latest react-native-permissions version, remove this dependency,
 since there is already a cross-platform openSettings method */
import { appDetailsSettings } from 'react-native-android-open-settings'
// TODO: using v1.2.1. When upgrading to RN60, use the latest version.
import Permissions, { Status } from 'react-native-permissions'
import { ScannerComponent } from '../component/scanner'
import { NoPermissionComponent } from '../component/noPermission'
import { ThunkDispatch } from '../../../store'
import { showErrorScreen } from '../../../actions/generic'
import { AppError } from '../../../lib/errors'
import { connect } from 'react-redux'
import ErrorCode from '../../../lib/errorCodes'
import { consumeInteractionToken } from '../../../actions/sso/consumeInteractionToken'

interface Props
  extends NavigationScreenProps,
    ReturnType<typeof mapDispatchToProps> {}

const CAMERA_PERMISSION = 'camera'

enum RESULTS {
  AUTHORIZED = 'authorized',
  RESTRICTED = 'restricted',
}

const IS_IOS = Platform.OS === 'ios'

export const ScannerContainer = (props: Props) => {
  const { consumeToken, showError, navigation } = props

  const [reRenderKey, setRenderKey] = useState(Date.now())
  const [permission, setPermission] = useState<Status>(RESULTS.RESTRICTED)
  const [isCameraReady, setCameraReady] = useState(false)
  const [isTorch, setTorch] = useState(false)
  const [isError, setError] = useState(false)
  const [colorAnimationValue] = useState(new Animated.Value(0))
  const [textAnimationValue] = useState(new Animated.Value(0))

  useEffect(() => {
    let focusListener
    if (navigation) {
      focusListener = navigation.addListener('willFocus', () => {
        // NOTE: the re-render and the re-mount should only fire during the willFocus event
        setRenderKey(Date.now())
      })
    }

    requestCameraPermission().then(() => {
      setTimeout(() => setCameraReady(true), 200)
    })

    return focusListener && focusListener.remove
  }, [])

  const requestCameraPermission = async () => {
    const permission = await Permissions.request(CAMERA_PERMISSION)
    setPermission(permission)
  }

  const openSettings = () => {
    const listener = async (state: AppStateStatus) => {
      if (state === 'active') {
        AppState.removeEventListener('change', listener)
        await requestCameraPermission()
      }
    }

    AppState.addEventListener('change', listener)

    try {
      const openPlatformSettings = Platform.select({
        ios: Permissions.openSettings,
        android: appDetailsSettings,
      })
      openPlatformSettings()
    } catch (e) {
      AppState.removeEventListener('change', listener)
    }
  }

  const onEnablePermission = async () => {
    if (IS_IOS) {
      openSettings()
    } else {
      if (permission === RESULTS.RESTRICTED) {
        openSettings()
      } else {
        await requestCameraPermission()
      }
    }
  }

  const animateColor = () =>
    Animated.sequence([
      Animated.timing(colorAnimationValue, {
        toValue: 1,
        duration: 300,
      }),
      Animated.timing(colorAnimationValue, {
        toValue: 0,
        delay: 400,
        duration: 300,
      }),
    ])

  const animateText = () =>
    Animated.sequence([
      Animated.timing(textAnimationValue, {
        toValue: 1,
        duration: 200,
      }),
      Animated.timing(textAnimationValue, {
        toValue: 0,
        delay: 1200,
        duration: 500,
      }),
    ])

  const startLocalNotification = () => {
    setError(true)
    Animated.parallel([animateColor(), animateText()]).start(() => {
      setError(false)
    })
  }

  const onScan = async (jwt: string) => {
    try {
      await consumeToken(jwt)
    } catch (e) {
      if (e instanceof SyntaxError) {
        startLocalNotification()
      } else if (e.message === 'Token expired') {
        showError(ErrorCode.TokenExpired, e)
      } else if (e.message === 'Signature on token is invalid') {
        showError(ErrorCode.InvalidSignature, e)
      } else if (
        e.message === 'You are not the intended audience of received token'
      ) {
        showError(ErrorCode.WrongDID, e)
      } else if (e.message === 'The token nonce does not match the request') {
        showError(ErrorCode.WrongNonce, e)
      } else {
        showError(ErrorCode.Unknown, e)
      }
    }
  }

  return permission === RESULTS.AUTHORIZED ? (
    isCameraReady ? (
      <ScannerComponent
        reRenderKey={reRenderKey}
        onScan={onScan}
        isTorchPressed={isTorch}
        onPressTorch={(state: boolean) => setTorch(state)}
        isError={isError}
        colorAnimationValue={colorAnimationValue}
        textAnimationValue={textAnimationValue}
      />
    ) : null
  ) : (
    <NoPermissionComponent onPressEnable={onEnablePermission} />
  )
}

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  consumeToken: (jwt: string) => dispatch(consumeInteractionToken(jwt)),
  showError: (errorCode: ErrorCode, e: Error) =>
    dispatch(showErrorScreen(new AppError(errorCode, e))),
})

export const Scanner = connect(
  null,
  mapDispatchToProps,
)(ScannerContainer)
