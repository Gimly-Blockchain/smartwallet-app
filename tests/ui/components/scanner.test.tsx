import React from 'react'
import { shallow } from 'enzyme'
import { ScannerComponent } from '../../../src/ui/interaction/component/scanner'
import { TouchableHighlight } from 'react-native'
import QRScanner, { Event } from 'react-native-qrcode-scanner'
import { stub } from '../../utils'

describe.only('QR Code component', () => {
  const props = {
    onScannerSuccess: jest.fn(),
  }

  //@ts-ignore
  const QRComponent = shallow(<ScannerComponent {...props} />)

  it('matches the snapshot on initial render', () => {
    expect(QRComponent).toMatchSnapshot()
  })

  it('correctly changes torch button while icon is touched', () => {
    // @ts-ignore
    QRComponent.find(TouchableHighlight)
      .props()
      .onPressIn()
    QRComponent.update()
    expect(QRComponent).toMatchSnapshot()
  })

  it('correctly changes torch button when icon is released', () => {
    // @ts-ignore
    QRComponent.find(TouchableHighlight)
      .props()
      .onPressOut()

    QRComponent.update()
    expect(QRComponent).toMatchSnapshot()
  })

  it('should correctly whatever', () => {
    const mockEvent = stub<Event>({ data: 'hello' })

    QRComponent.setState({ isCameraReady: true })
    QRComponent.find(QRScanner)
      .props()
      .onRead(mockEvent)

    expect(props.onScannerSuccess).toHaveBeenCalledWith('hello')
  })
})
