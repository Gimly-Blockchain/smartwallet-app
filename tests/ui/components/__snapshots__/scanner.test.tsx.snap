// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QR Code component correctly changes torch button when icon is released 1`] = `
<Fragment>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.65)",
          "height": 562.5,
          "width": 750,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "white",
          "borderRadius": 2,
          "borderWidth": 2,
          "height": 562.5,
          "justifyContent": "center",
          "width": 562.5,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.65)",
          "height": 562.5,
          "width": 750,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.65)",
        "flex": 1,
        "height": "100%",
        "justifyContent": "space-between",
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgb(255, 239, 223)",
          "flex": 1,
          "fontFamily": "TTCommons-Light",
          "fontSize": 20,
          "lineHeight": 22,
          "marginTop": 16,
          "textAlign": "center",
          "width": 562.5,
        }
      }
    >
      It's all automatic, just place your phone above the code
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <TouchableHighlight
        activeOpacity={1}
        delayPressOut={100}
        onPressIn={[Function]}
        onPressOut={[Function]}
        style={
          Object {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
            "width": 80,
          }
        }
        underlayColor="transparent"
      >
        <SvgTorchOffIcon />
      </TouchableHighlight>
    </View>
  </View>
</Fragment>
`;

exports[`QR Code component correctly changes torch button while icon is touched 1`] = `
<Fragment>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.65)",
          "height": 562.5,
          "width": 750,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "white",
          "borderRadius": 2,
          "borderWidth": 2,
          "height": 562.5,
          "justifyContent": "center",
          "width": 562.5,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.65)",
          "height": 562.5,
          "width": 750,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.65)",
        "flex": 1,
        "height": "100%",
        "justifyContent": "space-between",
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgb(255, 239, 223)",
          "flex": 1,
          "fontFamily": "TTCommons-Light",
          "fontSize": 20,
          "lineHeight": 22,
          "marginTop": 16,
          "textAlign": "center",
          "width": 562.5,
        }
      }
    >
      It's all automatic, just place your phone above the code
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <TouchableHighlight
        activeOpacity={1}
        delayPressOut={100}
        onPressIn={[Function]}
        onPressOut={[Function]}
        style={
          Object {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
            "width": 80,
          }
        }
        underlayColor="transparent"
      >
        <SvgTorchOnIcon />
      </TouchableHighlight>
    </View>
  </View>
</Fragment>
`;

exports[`QR Code component matches the snapshot on initial render 1`] = `
<Fragment>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.65)",
          "height": 562.5,
          "width": 750,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "white",
          "borderRadius": 2,
          "borderWidth": 2,
          "height": 562.5,
          "justifyContent": "center",
          "width": 562.5,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.65)",
          "height": 562.5,
          "width": 750,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.65)",
        "flex": 1,
        "height": "100%",
        "justifyContent": "space-between",
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgb(255, 239, 223)",
          "flex": 1,
          "fontFamily": "TTCommons-Light",
          "fontSize": 20,
          "lineHeight": 22,
          "marginTop": 16,
          "textAlign": "center",
          "width": 562.5,
        }
      }
    >
      It's all automatic, just place your phone above the code
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <TouchableHighlight
        activeOpacity={1}
        delayPressOut={100}
        onPressIn={[Function]}
        onPressOut={[Function]}
        style={
          Object {
            "alignItems": "center",
            "height": 80,
            "justifyContent": "center",
            "width": 80,
          }
        }
        underlayColor="transparent"
      >
        <SvgTorchOffIcon />
      </TouchableHighlight>
    </View>
  </View>
</Fragment>
`;
